math/src/main/java/com/powsybl/math/graph/UndirectedGraphImpl.java

```diff
@@ -543,14 +467,32 @@ public boolean traverse(int v, TraversalType traversalType, Traverser traverser,
         }
 
         TIntArrayList[] adjacencyList = getAdjacencyList();
-        boolean keepGoing;
+        boolean keepGoing = true;
 
-        if (traversalType == TraversalType.DEPTH_FIRST) {  // traversal by depth first
-            keepGoing = traverseDepthFirst(traverser, adjacencyList, encountered, v);
-        } else {  // traversal by breadth first
-            keepGoing = traverseBreadthFirst(traverser, adjacencyList, encountered, v);
-        }
+        encountered[v] = true;
+        Deque<Integer> edgesToTraverse = new ArrayDeque<>();
+        addEdges(edgesToTraverse, adjacencyList, v, traversalType);
+        while (!edgesToTraverse.isEmpty()) {
+            int e = switch (traversalType) {
+                case DEPTH_FIRST -> edgesToTraverse.pollLast();
+                case BREADTH_FIRST -> edgesToTraverse.pollFirst();
+            };
+            Edge<E> edge = edges.get(e);
+            if (encountered[edge.getV1()] && encountered[edge.getV2()]) {
+                continue; // traversing parallel edges only once
+            }
 
+            int vOrigin = encountered[edge.getV1()] ? edge.getV1() : edge.getV2();
+            int vDest = encountered[edge.getV1()] ? edge.getV2() : edge.getV1();
```

Comment:
Since it's the same condition, it might be easier to read (and it reduces the complexity) if you used an if statement, even if that means a bit more lines of code ?